engine/src/main/java/org/pentaho/di/trans/step/BaseStep.java

```diff
@@ -242,11 +241,15 @@
    */
   private List<RowSet> inputRowSets;
 
+  private final Object inputRowSetsLock = new Object();
```

Comment:
This would be much better using a ConcurrentReadWriteLock. Many of the accesses are read-only.

Reply:
@mbatchelor thanks for review.

I tried to avoid doing premature optimization. As far as I see, input/output row sets are read and modified mostly uncontended. When most of the time a single thread is accessing the collections, we would only increase execution cost with no benefits.

let me quote Brian Goetz
>In general, you should probably stick with synchronization unless you have a specific need for one of the advanced features of Lock, or if you have demonstrated evidence (not just a mere suspicion) that synchronization in this particular situation is a scalability bottleneck.
>- https://www.ibm.com/developerworks/java/library/j-jtp10264/

please, correct me if I'm wrong at my assumptions about contention (otherwise, we would have faced ConcurrentModificationException a lot earlier, wouldn't we?) and confirm we do need to convert synchronized to locks.

Thanks.